File: leetcode-algorithms/020. Valid Parentheses/solution.cpp
#include <iostream>
#include <map>
#include <stack>
#include <string>
using namespace std;


bool isValid(string s) {
	stack<char> st;
	map<char,char> m={  {")","("},
						{"}","{"},
						{"]","["}
					}

	for(auto c: s)
	{
		if(c=="["||c=="("||c=="{")
		{
			st.push(c);
			continue;
		}
		if(!st.empty() && m[c]==st.pop())
		{
			st.pop()
		}
		else
			return false;

	}
	return !st.empty();
}

int main()
{
	string s("()");
	bool res = isValid(s);
	cout<<res<<endl;
	return 0;
}

File: leetcode-algorithms/020. Valid Parentheses/solution.py
def isValid(s):
	"""
	:type s: str
	:rtype: bool
	"""

	if not s:
		return True

	satck = []

	k_v = {')':'(',']':'[','}':'{'}

	for c in s:
		if c in k_v.values():
			satck.append(c)
			continue
		if satck and k_v[c] == satck[-1]:
			satck.pop()
		else:
			return False

	return  not satck




if __name__ == '__main__':
	s=']'
	res = isValid(s)
	print(res)


